typescript/sdk/src/token/adapters/CosmosTokenAdapter.ts

```diff
@@ -38,28 +50,135 @@ export class NativeTokenAdapter
   }
 
   async populateTransferTx(
-    transferParams: TransferParams,
+    _transferParams: TransferParams,
+  ): Promise<MsgTransferEncodeObject> {
+    throw new Error('TODO not yet implemented');
+  }
+}
+
+export class CosmIbcTokenAdapter
+  extends CosmNativeTokenAdapter
+  implements IHypTokenAdapter
+{
+  constructor(
+    public readonly chainName: ChainName,
+    public readonly multiProvider: MultiProtocolProvider,
+    public readonly addresses: Record<string, Address>,
+    public readonly properties: {
+      ibcDenom: string;
+      sourcePort: string;
+      sourceChannel: string;
+    },
+  ) {
+    if (
+      !properties.ibcDenom ||
+      !properties.sourcePort ||
+      !properties.sourceChannel
+    )
+      throw new Error('Missing properties for CosmNativeIbcTokenAdapter');
+    super(chainName, multiProvider, addresses, properties);
+  }
+
+  getDomains(): Promise<Domain[]> {
+    throw new Error('Method not applicable to IBC adapters');
+  }
+  getRouterAddress(_domain: Domain): Promise<Buffer> {
+    throw new Error('Method not applicable to IBC adapters');
+  }
+  getAllRouters(): Promise<
+    Array<{
+      domain: Domain;
+      address: Buffer;
+    }>
+  > {
+    throw new Error('Method not applicable to IBC adapters');
+  }
+  quoteGasPayment(_destination: Domain): Promise<string> {
+    throw new Error('Method not applicable to IBC adapters');
+  }
+
+  async populateTransferRemoteTx(
+    transferParams: TransferRemoteParams,
+    memo = '',
   ): Promise<MsgTransferEncodeObject> {
-    const transfer: MsgTransfer = {
-      sourcePort: '',
-      sourceChannel: '',
+    if (!transferParams.fromAccountOwner)
+      throw new Error('fromAccountOwner is required for ibc transfers');
+
+    const value = {
+      sourcePort: this.properties.sourcePort,
+      sourceChannel: this.properties.sourceChannel,
       token: {
-        denom: this.ibcDenom,
+        denom: this.properties.ibcDenom,
         amount: transferParams.weiAmountOrId.toString(),
       },
-      sender: '',
-      receiver: '',
-      timeoutHeight: {
-        revisionNumber: 0n,
-        revisionHeight: 0n,
-      },
-      timeoutTimestamp: 0n,
-      memo: '', // how to encode this?
+      sender: transferParams.fromAccountOwner,
+      receiver: transferParams.recipient,
+      // Represented as nano-seconds
+      timeoutTimestamp: Long.fromNumber(
+        new Date().getTime() + COSMOS_IBC_TRANSFER_TIMEOUT,
+      ).multiply(1_000_000),
+      memo,
     };
     return {
       typeUrl: '/ibc.applications.transfer.v1.MsgTransfer',
-      // @ts-ignore
-      value: transfer,
+      value,
+    };
+  }
+}
+
+export class CosmIbcToWarpTokenAdapter
```

Comment:
I don't quite understand the class hierachy here, is `CosmIbcTokenAdapter` used?

Reply:
Oh for just plain IBC transfers? Some kind of comment to that effect would be helpful

Reply:
Correct.

Reply:
Added comments